hax/hax/util.py

```diff
@@ -589,6 +589,62 @@ def get_node_health(self, node: str) -> str:
             raise HAConsistencyException(
                 f'Failed to get {node} node health') from e
 
+    def get_node_fid(self, node: str) -> Optional[str]:
+        # Example,
+        # m0conf/nodes/
+        # 0x6e00000000000001:0x3:{"name": "ssc-vm-1623.colo.seagate.com",
+        #                         "state": "M0_NC_UNKNOWN"}
+        node_items = self.kv.kv_get('m0conf/nodes', recurse=True)
+        for item in node_items:
+            item_value = json.loads(item['Value'])
+            if 'name' in item_value and item_value['name'] == node:
+                return str(item['Key'].split('/')[2])
+        return None
+
+    def get_node_ctrl_fid(self, node: str) -> Optional[str]:
+        # Example,
+        # {
+        #    "key": "m0conf/sites/0x5300000000000001:0x1/
+        #            racks/0x6100000000000001:0x2/encls/
+        #            0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5",
+        # },
+        encl_fid = self.get_node_encl_fid(node)
+        if not encl_fid:
+            return None
+        ctrl_items = self.kv.kv_get('m0conf/sites', recurse=True)
+        regex = re.compile(
+            f'^m0conf\\/.*\\/racks\\/.*\\/encls\\/{encl_fid}\\/ctrls\\/(.+)$')
+        for ctrl in ctrl_items:
+            match_result = re.match(regex, ctrl['Key'])
+            if not match_result:
+                continue
+            return str(ctrl['Key'].split('/')[8])
+        return None
+
+    def get_node_encl_fid(self, node: str) -> Optional[str]:
+        # Example,
+        # {
+        #    "key": "m0conf/sites/0x5300000000000001:0x1/
+        #            racks/0x6100000000000001:0x2/encls/
+        #            0x6500000000000001:0x4",
+        #    "value": "{\"node\": \"0x6e00000000000001:0x3\",
+        #               \"state\": \"M0_NC_UNKNOWN\"}"
+        # },
+        node_fid = self.get_node_fid(node)
+        if not node_fid:
+            return None
+        encl_items = self.kv.kv_get('m0conf/sites', recurse=True)
+        regex = re.compile(
+            '^m0conf\\/.*\\/racks\\/.*\\/encls\\/(.+)$')
```

Comment:
According to 4/KV RFC, there are two possibilities:
```
`m0conf/sites/<site_fid>/racks/<rack_fid>/encls/<encl_fid>` | `{ "state": "<HA state>" }` | [HA state](#ha-state) of this enclosure.
`m0conf/sites/<site_fid>/racks/<rack_fid>/encls/<encl_fid>/ctrls/<ctrl_fid>` | `{ "node": "<node_fid>", "state": "<HA state>" }` | Fid of the corresponding node and [HA state](#ha-state) of this controller.
```
Both of the keys will match your regexp but in first case matched group will contain FID only, while the second will contain "<FID>/ctrls/<FID>". Is it expected by your code?

Reply:
Oh!!, my requirement is for matching only first type.

<!--StartFragment-->
m0conf/sites/&lt;site_fid&gt;/racks/&lt;rack_fid&gt;/encls/&lt;encl_fid&gt; | { "node": "&lt;node_fid&gt;", "state": "&lt;HA state&gt;" }
-- | --

<!--EndFragment-->

I got a better regex matching only this type.
` regex=re.compile('^m0conf\\/.*\\/racks\\/.*\\/encls\\/[^/]+$')`

Shall I update it?

Reply:
Yes, please

Reply:
done